hazelcast/src/test/java/com/hazelcast/cluster/TcpIpJoinTest.java

```diff
@@ -224,4 +232,103 @@ public void test_whenIncompatibleJoiners() throws Exception {
+            @Override
+            public void run() {
+                assertContainsAll(tcpJoiner1.getKnownMemberAddresses().keySet(), Arrays.asList(hz2Address, hz3Address));
+            }
+        }, 15);
+    }
+
+    @Test
+    public void test_tcpIpJoinerCleanupAddressesAfterAddressRetentionPeriodIsPassed() throws Exception {
+        Config config1 = smallInstanceConfig()
+                .setProperty("hazelcast.merge.first.run.delay.seconds", "10")
+                .setProperty("hazelcast.merge.next.run.delay.seconds", "10");
+        JoinConfig join1 = config1.getNetworkConfig().getJoin();
+        join1.getMulticastConfig().setEnabled(false);
+        TcpIpConfig tcpIpConfig1 = join1.getTcpIpConfig();
+        tcpIpConfig1.setEnabled(true);
+
+        Config config2 = smallInstanceConfig();
+        JoinConfig join2 = config2.getNetworkConfig().setPort(8899).getJoin();
+        join2.getMulticastConfig().setEnabled(false);
+        TcpIpConfig tcpIpConfig2 = join2.getTcpIpConfig();
+        tcpIpConfig2.setEnabled(true).addMember("127.0.0.1:5701");
+
+        Config config3 = smallInstanceConfig();
+        JoinConfig join3 = config3.getNetworkConfig().setPort(9099).getJoin();
+        join3.getMulticastConfig().setEnabled(false);
+        TcpIpConfig tcpIpConfig3 = join3.getTcpIpConfig();
+        tcpIpConfig3.setEnabled(true).addMember("127.0.0.1:5701");
+
+        HazelcastInstance hz1 = Hazelcast.newHazelcastInstance(config1);
+        TcpIpJoiner tcpJoiner1 = (TcpIpJoiner) TestUtil.getNode(hz1).getJoiner();
+        overrideAddressRetentionDuration(tcpJoiner1, TimeUnit.SECONDS.toMillis(10));
+
+        HazelcastInstance hz2 = Hazelcast.newHazelcastInstance(config2);
+        HazelcastInstance hz3 = Hazelcast.newHazelcastInstance(config3);
+        Address hz2Address = hz2.getCluster().getLocalMember().getAddress();
+        Address hz3Address = hz3.getCluster().getLocalMember().getAddress();
+        assertTrueEventually(new AssertTask() {
+            @Override
+            public void run() {
+                assertContainsAll(tcpJoiner1.getKnownMemberAddresses().keySet(), Arrays.asList(hz2Address, hz3Address));
+            }
+        });
+        hz3.shutdown();
+        // Timeout below must be greater than both hazelcast.merge.next.run.delay.seconds and previouslyJoinedMemberAddressRetentionDuration
+        assertTrueEventually(new AssertTask() {
+            @Override
+            public void run() {
+                Set<Address> rememberedAddresses = tcpJoiner1.getKnownMemberAddresses().keySet();
+                assertContains(rememberedAddresses, hz2Address);
+                assertNotContains(rememberedAddresses, hz3Address);
+            }
+        }, 20);
+    }
+
+    private static void overrideAddressRetentionDuration(TcpIpJoiner tcpIpJoiner, long addressRetentionDuration) throws Exception {
+        Field privateField = TcpIpJoiner.class.getDeclaredField("previouslyJoinedMemberAddressRetentionDuration");
```

Comment:
Why not make field package-private with `@VisibleForTesting`?

Reply:
I also want to keep this field final. I know this reflection usage doesn't look good but I want it to be immutable. If you prefer avoiding that reflection, I can change it to your preferred way.